### PR TITLE
Make default time at 00 00 instead of 12 00

### LINKED ISSUE #9989: Make default time at 00:00 instead of 12:00
## Context
Currently when using reports, the default time is 12:00 so users need to constantly change to 00:00 to not miss any info.

The date selector is used on multiple pages:
- Reports
- BOM (Bulk Order Management)
- Orders
- Create OC
- maybe others?

It would be good to use this issue to bring consistency across the Date selectors.

## Description
The default start and end time for Reports, BOM and Orders
Start Time >= 00:00
End Time < 00:00

The default time in Order Cycles should not change if possible (stay at 12:00) to minimise potential for hub managers to make a mistake with this change.

You should be able to set the time in all places the date selector is used.

## Acceptance Criteria

- Setting start and end time should be possible in all places date selector is used
- Times set by default are as stated in the description.
- The filters capture all orders of these times. It will be tricky to test the edge cases around time of order placement so need to be sure that no orders will be missed.

### PR DESCRIPTION
**Pull request by abdellani**
#### What? Why?

- Closes #9989 

#### What should we test?

- described on #9989

#### Release notes

Changelog Category: User facing changes

### CODE REVIEW

**Comment by abdellani** [comment]
The failing test doesn't seem to be related to the changes.

File: app/webpacker/controllers/flatpickr_controller.js [hunk 1]
```diff
@@ -60,6 +60,14 @@ export default class extends Flatpickr {
 
   open() {
     this.fp.element.dispatchEvent(new Event("focus"));
+    if (!this.defaultDateValue) return;
+    if (this.fp.selectedDates.length) return;
```
Comment (abdellani) [inline app/webpacker/controllers/flatpickr_controller.js:64]: if a date is already selected, we don't need to set a default value

File: app/webpacker/controllers/flatpickr_controller.js [hunk 2]
```diff
@@ -60,6 +60,14 @@ export default class extends Flatpickr {
...
+
+    if (this.defaultDateValue === "startOfDay") {
+      this.fp.setDate(moment().startOf("day").format());
+    } else if (this.defaultDateValue === "endOfDay") {
+      this.fp.setDate(moment().add(1, "days").startOf("day").format());
```
Comment (jibees) [inline app/webpacker/controllers/flatpickr_controller.js:69]: I'm wondering when you didn't use `moment().endOf('day')`?

https://momentjscom.readthedocs.io/en/latest/moment/03-manipulating/04-end-of/

File: app/webpacker/controllers/flatpickr_controller.js [hunk 3]
```diff
@@ -17,7 +17,7 @@ import ShortcutButtonsPlugin from "shortcut-buttons-flatpickr";
 import labelPlugin from "flatpickr/dist/plugins/labelPlugin/labelPlugin";
 
 export default class extends Flatpickr {
-  static values = { enableTime: Boolean, mode: String };
+  static values = { enableTime: Boolean, mode: String, defaultDate: String };
```
Comment (jibees) [inline app/webpacker/controllers/flatpickr_controller.js:20]: I think it's great to indicates as a comment that `defaultDate` only takes `startOfDay` and `endOfDay` as accepted values. What do you think?

File: app/webpacker/controllers/flatpickr_controller.js [hunk 4]
```diff
@@ -60,6 +60,14 @@ export default class extends Flatpickr {
 
   open() {
     this.fp.element.dispatchEvent(new Event("focus"));
+    if (!this.defaultDateValue) return;
```
Comment (jibees) [inline app/webpacker/controllers/flatpickr_controller.js:63]: I've another suggestion, that is, to me, easier to understand:
```js
  open() {
    this.fp.element.dispatchEvent(new Event("focus"));
    if (!this.fp.selectedDates.length) {
      this.setDefaultDateValue();
    }
  }

  ....
  // private
  setDefaultDateValue = () => {
    if (this.defaultDateValue === "startOfDay") {
      this.fp.setDate(moment().startOf("day").format());
    } else if (this.defaultDateValue === "endOfDay") {
      this.fp.setDate(moment().add(1, "days").startOf("day").format());
    }
  };
```
(not tested)
What do you think?

File: app/webpacker/controllers/flatpickr_controller.js (hunk 1 above)
Comment (jibees) [reply app/webpacker/controllers/flatpickr_controller.js:64]: 👌

**Review by jibees — COMMENTED** [review]
Thanks for this one!
I've few suggestions, let me know if it's ok to you ;)

File: app/webpacker/controllers/flatpickr_controller.js (hunk 2 above)
Comment (abdellani) [reply app/webpacker/controllers/flatpickr_controller.js:69]: Hi @jibees 
I didn't use that function because the form ignores the seconds part from the date.
So, the report will not load the data between [23:59:00 ~> 23:59:59]
image
What do you think?

File: app/webpacker/controllers/flatpickr_controller.js (hunk 3 above)
Comment (abdellani) [reply app/webpacker/controllers/flatpickr_controller.js:20]: I like the idea

File: app/webpacker/controllers/flatpickr_controller.js (hunk 4 above)
Comment (abdellani) [reply app/webpacker/controllers/flatpickr_controller.js:63]: I agree.
I tried to do that but the code was crashing because I didn't use `this` before `setDefaultDateValue`.

I'll follow this recommendation :)

File: app/webpacker/controllers/flatpickr_controller.js (hunk 2 above)
Comment (jibees) [reply app/webpacker/controllers/flatpickr_controller.js]: That's a very edge case which you took into account, thanks!
I'd vote for a small comment in the code that explain that. What do you think?

File: app/webpacker/controllers/flatpickr_controller.js [hunk 5]
```diff
@@ -60,6 +63,10 @@ export default class extends Flatpickr {
...
```
Comment (jibees) [inline app/webpacker/controllers/flatpickr_controller.js:66]: I'd delete this line to which seems to me useless ; or maybe as the first line in `setDefaultDateValue()` method?

File: app/webpacker/controllers/flatpickr_controller.js [hunk 6]
```diff
@@ -136,4 +143,12 @@ export default class extends Flatpickr {
         break;
     }
   };
+
+  setDefaultDateValue() {
+    if (this.defaultDateValue === "startOfDay") {
+      this.fp.setDate(moment().startOf("day").format());
+    } else if (this.defaultDateValue === "endOfDay") {
+      this.fp.setDate(moment().add(1, "days").startOf("day").format());
```
Comment (jibees) [inline app/webpacker/controllers/flatpickr_controller.js:151]: This probably needs a small comment on that `23:59:00` vs. `23:59:59` thing ;)

**Review by jibees — COMMENTED** [review]
Just add some comments. Otherwise, it's good to me!

File: app/webpacker/controllers/flatpickr_controller.js (hunk 6 above)
Comment (abdellani) [reply app/webpacker/controllers/flatpickr_controller.js:154]: done : )

**Review by jibees — APPROVED** [review]
LGTM!

Thanks and sorry for the back and forth!

**Comment by drummer83** [comment]
Hi @abdellani,
And thanks for working on this!

Here are the places which I found with a date/time selector:
| page | type | default BEFORE | default AFTER |
|-|-|-|-|
| /admin/order_cycles | open & close | no date, 12:00 | no date, 12:00 |
| /admin/order_cycles/new | open & close | no date, 12:00 | no date, 12:00 |
| /admin/order_cycles/1234/edit | open & close | no date, 12:00 | no date, 12:00 |
| /admin/reports/ with no pre-filled date/time range | start & end | OnClick: no date, 12:00 | OnClick: today, 00:00 & tomorrow, 00:00 :exclamation: | 
| /admin/reports/ with pre-filled date/time range | start & end | Pre-filled: past month, 00:00 & tomorrow, 00:00 | Pre-filled: past month, 00:00 & tomorrow 00:00 | 

No pre-filled date range:
- Orders And Distributors
- Bulk Co-op Supplier Report
- Bulk Co-op Allocation
- Bulk Co-op Packing Sheets
- Bulk Co-op Customer Payments
- Payments By Type
- Itemised Payment Totals
- Payment Totals
- Enterprise Fee Summary
- Tax Types
- Tax Rates
- Summary
- Detailed

Pre-filled date range:
- Order Cycle Supplier Totals
- Order Cycle Supplier Totals by Distributor
- Order Cycle Distributor Totals by Supplier
- Order Cycle Customer Totals
- Payment Methods Report
- Delivery Report
- Pack By Customer
- Pack By Supplier
- Pack By Product
- Revenues by Hub

These places have date selectors only:
| page | type | default BEFORE | default AFTER |
|-|-|-|-|
| /admin/subscriptions/12/edit | begins at | previous value | previous value |
| /admin/subscriptions/12/edit | ends at | previous value or no date | previous value or no date |
| /admin/subscriptions/new | begins at | no date | no date |
| /admin/subscriptions/new | ends at | no date | no date |

Here go the date range selectors:
| page | type | default BEFORE | default AFTER |
|-|-|-|-|
| /admin/orders | date range | no date range | no date range |
| /admin/orders/bulk_management | date range | past week | past week |

## Results
- We can see that only for reports without a pre-filled date/time range the time has been changed to 00:00 for start and end dates. :heavy_check_mark: 
- All other occurrences of the date, date/time or date range selectors remain unchanged. :heavy_check_mark: 
  - Not so sure if this is enough to fulfill the request to be more consistent within reports by @RachL or @lin-d-hop in the [original issue](https://github.com/openfoodfoundation/openfoodnetwork/issues/9989#issue-1447994548) ('It would be good to use this issue to bring consistency across the Date selectors.') If more is needed, we need to do this in a separate PR.
- The date/time filters are filtering correctly, as can be seen in this example:
image

## Conclusion
I think this one is ready to go although I am not 100 % convinced about the usability as I tried to explain in [this comment](https://github.com/openfoodfoundation/openfoodnetwork/issues/9989#issuecomment-1313724969). We will see if other people feel the same and we will need another iteration here.

Thanks again Mohamed!

**Comment by abdellani** [comment]
Hi @drummer83 
Thank you for your feedback.

When I set the default end date to [Tomorrow 00:00] instead of [Today 23:59], I was thinking about an edge case where there are transactions between [Today 23:59:**00** ~> Today 23:59:**59**] (our data selector ignores the seconds).

With this update, if the user selects 7th to 13th Nov, we'll get the transaction [7th Nov 00:00 ~> 13th Nov 00:00].
If the user doesn't notice the default hour (Which I think can be common), he'll not see the transactions happening in the range [13th Nov 00:00:**00** ~ 13th Nov 23:59:**59**]

After thinking again, I think it'll be better to miss 1 min rather than 24h.

**Comment by drummer83** [comment]
Hi @abdellani and @openfoodfoundation/train-drivers-product-owners,
I don't want to block anything here just pointing at an issue or complaint I am expecting to come. Happy to merge as is if this is what you prefer.

Missing orders is never good, no matter if 1 minute or 24 hours. Would it be possible to show 'today 23:59` to the user and use' tomorrow 00:00' in the query?

I am hesitant about usability. Maybe I will ask some instance managers for their opinion on Slack.

**Comment by RachL** [comment]
> if the user selects 7th to 13th Nov, we'll get the transaction [7th Nov 00:00 ~> 13th Nov 00:00]

To me this looks better than what we have today, which is 7th Nov 12:00 -> 13th Nov 12:00. Currently what users do is that they manually change for 00:00. This saves them the time to do this change manually, which was the purpose of this issue. 

Ideally, in the future we will have different type of selectors : past months, past year, past week etc and we won't have to over think default dates. But that's a bigger work. Here we just wanted to focus on removing the 12:00 problem in the display. So it looks like this is achieved and we can already merge this?
